### PR TITLE
From Docplex to Ising Model (QUBO)

### LINKED ISSUE #36: Transform a docplex optimization problem into openQAOA Hamiltonian
---
name: [Docplex](https://pypi.org/project/docplex/) converter.
about: Having an optimization problem in Docplex form transforms it into the OpenQAOA Hamiltonian.
title: 'docplex2hamiltonian'
labels: 'feature request'
assignees: @alejomonbar @MaldoAlberto 

---

### Prerequisites
Before raising this feature request, I have made sure:

- [x] It does not lead to duplication.
- [x] It has not already been requested.

### Is your feature request related to a problem? Please describe.

Implementing optimization problems using openQAOA is restricted to certain problems with a predefined structure ( NumberPartition, TSP, Knapsack, etc). For example, if we want to modify the Knapsack problem with a Heuristic equation, the implemented Knapsack problem would not help me to solve my problem and I must create my own Hamiltonian of this problem. Docplex a python based library used to solve optimization problems can be used to codify the optimization problem and then translate it directly into its OpenQAOA Hamiltonian.

It will be something like:

``` python
from docplex.mp.model import Model
from openqaoa.utilities import X_mixer_hamiltonian
from openqaoa.problems.helper_functions import docplex2hamiltonian #The function we want to implement

mdl = Model(name="MyOptimizationProblem") # Docplex model

z = mdl.binary_var_list(10, name="Z") #My binary optimization variables


objective = mdl.sum(z) - 2 * z[0]  + z[3] * z[4]  # Cost function

mdl.minimize(objective) # Adding the objective function to the model 

cost_hamil = docplex2hamiltonian(mdl) # The function we want to implement

mixer_hamil = X_mixer_hamiltonian(num_assets)

circuit_params = QAOACircuitParams(cost_hamil, mixer_hamil, p=1)


variate_params = create_qaoa_variational_params(circuit_params, 'standard', 'ramp')
backend_obj_qiskit = get_qaoa_backend(circuit_params, 'qiskit_shot_simulator', n_shots=10000)
backend_obj_qiskit.expectation_w_uncertainty(variate_params)
backend_obj_qiskit.expectation(variate_params)
optimizer_dict = {'method': 'cobyla', 'maxiter': 10}
variate_params_qiskit = variate_params
optimizer_obj_qiskit = get_optimizer(backend_obj_qiskit, variate_params_qiskit, optimizer_dict)
optimizer_obj_qiskit()
optimizer_obj_qiskit.results_information()
```

### Describe the solution you'd like
Create a class function that translates an optimization problem into its Hamiltonian representation.

### Describe alternatives you've considered
we think Numba can be used instead of Docplex. But, we should start first with Docplex.

### Additional context
Screen Shot 2022-06-17 at 8 07 02 AM
Screen Shot 2022-06-17 at 8 06 52 AM

Screen Shot 2022-06-17 at 7 30 50 AM

### PR DESCRIPTION
**Pull request by alejomonbar**
Adding functionality to translate docplex problems into its QUBO representation for OpenQAOA. This function allows to write objective functions with their respective equality and inequality constraints and transform them into its QUBO representation. They corresponds to #36 and #41.

## Description

In #36 and #41 there are explanations of this approach.
- Please also include relevant context.
- List any dependencies that are required for this change, if any. `docplex`
- **Fixes #36**

## Checklist

[//]: <> (- [x] My code follows the style guidelines of this project)

- [x] I have performed a self-review of my code.
- [x] I have commented my code and used numpy-style docstrings
- [ ] I have made corresponding updates to the documentation.
- [x] My changes generate no new warnings
- [x] I have added/updated tests to make sure bugfix/feature works.
- [x] New and existing unit tests pass locally with my changes.

[//]: <> (- [ ] Any dependent changes have been merged and published in downstream modules)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

`test_converters.py`

### CODE REVIEW

**Comment by Q-lds** [comment]
Hi @alejomonbar! What version of docplex are you using? We'll need to add it to the setup file, otherwise the tests won't run.

**Comment by alejomonbar** [comment]
> Hi @alejomonbar! What version of docplex are you using? We'll need to add it to the setup file, otherwise the tests won't run.

Hi @Q-lds, I'm using docplex version 2.23.222

File: openqaoa/problems/converters.py [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+        # considering two index i and j
+        for x, y, weight in expr.iter_quad_triplets():
+            i = self.idx_terms[x]
+            j = self.idx_terms[y]
+            if i == j:  # if this is the term is  Z1**2 for example
+                self.qubo_dict[(i,)] += weight
+            else:
+                self.qubo_dict[(i, j)] += weight
+
+    def Equality2Penalty(self, expression, multiplier: float):
```
Comment (vishal-ph) [inline openqaoa/problems/converters.py:80]: Why did we choose the camelcase for a class method here?

Comment (vishal-ph) [reply openqaoa/problems/converters.py:80]: Also, perhaps this method should be a `staticmethod`? I don't see any updates to `self`

File: openqaoa/problems/converters.py [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        Returns
+        -------
+        penalty : docplex.mp.quad.QuadExpr
+            Penalty that will be added to the cost function.
+
+        """
+        penalty = multiplier * (expression) ** 2
+        return penalty
+
+    def bounds(self, linear_expression):
```
Comment (vishal-ph) [inline openqaoa/problems/converters.py:101]: This method also seems like it could be a staticmethod

File: openqaoa/problems/converters.py [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        -------
+        l_bound : float
+            Lower bound of the quadratic term.
+        u_bound : float
+            Upper bound of the quadratic term
+        """
+
+        l_bound = (
+            u_bound
+        ) = linear_expression.constant  # Lower and upper bound of the contraint
```
Comment (vishal-ph) [inline openqaoa/problems/converters.py:120]: typo in constraint spelling

File: openqaoa/problems/converters.py [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+        """
+        l_bound = 0
+        u_bound = 0
+        for z1, z2, coeff in iter_exp:
+            l_bound += min(0, coeff)
+            u_bound += max(0, coeff)
+        return l_bound, u_bound
+
+    def Inequaility2Equality(self, constraint):
```
Comment (vishal-ph) [inline openqaoa/problems/converters.py:151]: CamelCase for class method

File: openqaoa/problems/converters.py (hunk 3 above)
Comment (alejomonbar) [reply openqaoa/problems/converters.py:120]: done

File: openqaoa/problems/converters.py (hunk 2 above)
Comment (alejomonbar) [reply openqaoa/problems/converters.py:101]: sure, I'll fix it.

File: openqaoa/problems/converters.py (hunk 1 above)
Comment (alejomonbar) [reply openqaoa/problems/converters.py:80]: I will modify the camelcase and the static method, thanks Vishal! :)

Comment (vishal-ph) [reply openqaoa/problems/converters.py:80]: great, thanks!

**Review by vishal-ph — APPROVED** [review]

**Pull request merged**